lib/model/compose.dart

```diff
@@ -0,0 +1,108 @@
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedOpeningBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length + 1);
+  }
+  return result;
+}
+
+/// Wrap Markdown [content] with opening and closing backtick fences.
+///
+/// For example:
+///
+///   const str =
+///     '```javascript\n'
+///     'console.log(\'Hello world!\');\n'
+///     '```';
+///   print(wrapWithBacktickFence(content, 'quote'));
+///   // Output:
+///   //
+///   // ````quote
+///   // ```javascript
+///   // console.log('Hello world!');
+///   // ```
+///   // ````
+///
+/// This does not parse [content] to make sure its backtick fences are properly
+/// paired and nested. If they aren't, it's probably not clear what
+/// backtick-fenced output would be more reasonable anyway --
+/// especially for callers like quote-and-reply that use this string as a
+/// best-effort suggestion for the user to inspect and fix before sending.
+/// (Render previews, #178, will help the user with that.)
+///
+/// In [content], indented code blocks
+///   ( https://spec.commonmark.org/0.30/#indented-code-blocks )
+/// and code blocks fenced with tildes should make no difference to the
+/// backtick fences we choose here; this function ignores them.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+// TODO(#178) Remove mention of #178 in doc.
+String wrapWithBacktickFence({required String content, String? infoString}) {
+  assert(infoString == null || !infoString.contains('`'));
+  assert(infoString == null || infoString.trim() == infoString);
+
+  StringBuffer resultBuffer = StringBuffer();
+
+  // (A) Why not panic on dangling opening fences
+  //     (ones without a corresponding closing fence)?
+  //     Because the spec allows leaving the closing fence implicit:
+  //     > If the end of the containing block (or document) is reached
+  //     > and no closing code fence has been found,
+  //     > the code block contains all of the lines after the opening code fence
+  //     > until the end of the containing block (or document).
+  //
+  // (B) Why not look for dangling closing fences (ones without an opening fence)?
+  //     Because technically there's no such thing:
+  //     they would be indistinguishable from dangling opening fences,
+  //     and parsers will treat them that way. (See A for what that treatment is.)
+  final fenceLength = getUnusedOpeningBacktickFenceLength(content);
+
+  resultBuffer.write('`' * fenceLength);
+  if (infoString != null) {
+    resultBuffer.write(infoString);
+  }
+  resultBuffer.write('\n');
+  resultBuffer.write(content);
+  resultBuffer.write('\n');
+  resultBuffer.write('`' * fenceLength);
+  return resultBuffer.toString();
```

Comment:
I think this should be writing a newline at the end, too.

That's necessary in order for the string to consist of complete lines, rather than end with a line fragment. And the syntax we're aiming to produce is a multi-line piece of syntax, in the context of Markdown which is quite line-oriented in general.

Concretely, if we take the output of this and it lacks a final newline, and then go on to put anything at all after it without first adding that missing newline, then that subsequent text will not be interpreted as intended, and will corrupt the closing code fence that this function intended. That's because that subsequent text (or the beginning of it) would become part of the last line this function started.